### ISSUE TITLE
LocalDate field in Record is not properly populated by RFC3339 date string

### ISSUE THREAD

**Issue by alf** [jOOQ/jOOQ#12798]
### Expected behavior

Given a record with a field of type LocalDate, I expect to be able to populate
that field with an RFC3339 date string. Unfortunately, the following test fails:

```
    @Test
    void woot() {
        var rec = new UtvekslingspersonRecord();
        var field = rec.field("DATO_FRA");
        rec.from(new String[] {"2020-12-31"}, field);
        assertNotNull(rec.get(field));
    }
```

### Actual behavior

jOOQ expects an RFC3339 date-time string. In other words, the following test succeeds:

```
    @Test
    void woot() {
        var rec = new UtvekslingspersonRecord();
        var field = rec.field("DATO_FRA");
        rec.from(new String[] {"2020-12-31T00:00"}, field);
        assertNotNull(rec.get(field));
    }
```

### Steps to reproduce the problem

This problem seems to be caused by the fact that we have enabled the dateAsTimestamp setting and then use a converter for fields we know only contains the date part:

```
...
     <dateAsTimestamp>true</dateAsTimestamp>
...
         <forcedtype>
             <userType>java.time.LocalDate</userType>
             <includeExpression>.*DATO_FRA</includeExpression>
             <converter>
                 org.jooq.Converter.ofNullable(java.time.LocalDateTime.class, LocalDate.class, java.time.LocalDateTime::toLocalDate, LocalDate::atStartOfDay)
             </converter>
         </forcedtype>
...
```

Our configuration generates the following field-definition:

```
    public final TableField<UtvekslingspersonRecord, LocalDate> DATO_FRA = createField(DSL.name("DATO_FRA"), SQLDataType.LOCALDATETIME(0).nullable(false), this, "Dato utvekslingen starter...", org.jooq.Converter.ofNullable(java.time.LocalDateTime.class, LocalDate.class, java.time.LocalDateTime::toLocalDate, LocalDate::atStartOfDay), new LocalDateAsLocalDateTimeBinding());
```

The following test seems to demonstrates the issue a bit more directly:

```
    @Test
    void woot() {
        var rec = new UtvekslingspersonRecord();
        var field = rec.field("DATO_FRA");
        var dt = field.getDataType();
        var v = dt.convert("2020-12-31");
        assertNotNull(v);
    }
```

Inspecting `dt` shows that this is a ConvertedDataType of type LocalDate with a
DefaultDataType delegate of type LocalDateTime. The problem seems to be that
`dt.convert` calls `delegate.convert`, and thus ends up converting the object
using the wrong type.

### Versions

- jOOQ: 3.15.5
- Java: openjdk version "11.0.12" 2021-07-20
- Database (include vendor): Oracle XE 18c
- OS: Linux
- JDBC Driver (include name if inofficial driver): oracle ojdbc8 18.4.0.0_181016

**Comment by lukaseder** [comment]
Thanks a lot for your report. Your `Converter<T, U>` binds `T = LocalDateTime` and `U = LocalDate`. A priori, jOOQ has no knowledge about `U` types (where `U` stands for user type), so it doesn't really know how to convert any arbitrary type (e.g. `String` to `U`). But it does know how to convert `String` to `T`.

Think of the `dt.convert(Object)` method as a `Converter<T, Object>` (in reverse) chained to a `Converter<T, U>`, so jOOQ tries to convert the chain `Object -> T -> U`.

I can see how it may look like intuitively, jOOQ should also know how to convert `Object` to `U`, because `U` is such a common type in your case. But formally, I don't think there's a simple answer to find out how to do that. After all, we can't just bypass your own converter. What if your converter did some time zone mangling, and the correct way to interpret `2020-12-31T23:00:00` would actually be to produce `2021-01-01`? If your converter would be bypassed, the result would be simply wrong.

I hope this makes sense?

**Comment by alf** [comment]
I’m not sure I follow. I would have expected jOOQ to call the converters in reverse. That is:
`dt.convert(Object)` would do `Object -> U -> T`.

I accept that this is not a bug but rather a user error. I hope you can assist me in finding a solution. The reason this problem exists, is that I have enabled dateAsTimestamp. Ideally I’d have preferred to only do this for a subset of my fields, but I haven’t found a way to accomplish that.

Some of our DATE-fields contains a time component, others don’t. We can divine which are which using our data catalog so in theory I should be able to handle this using jOOQ and thus spare our developers from having to know.

The above is the closest I’ve come, but as you see it breaks down when parsing the incoming user object.

**Comment by lukaseder** [comment]
OK, so another example. Your object `U` is:

```java
record MyRecord(int a, String b, RuntimeException c) {}
```

You have found a useful way to map `T = String` to the above `U` type. Now you want to convert some value of some type using `dt.convert(myValue)`. How would jOOQ convert from `Object -> U`?

It can't work. In your case, it *looks as though* it might work, but it can't, because it isn't really a different case. There's no way jOOQ can guess that when going from `String` to `U`, the "direct approach" (e.g. `LocalDate::parse`) is still correct, i.e. that `LocalDate::atStartOfDay` isn't relevant in this case.

> I accept that this is not a bug but rather a user error. I hope you can assist me in finding a solution.

I knew I should have asked for the real problem :)

> Ideally I’d have preferred to only do this for a subset of my fields, but I haven’t found a way to accomplish that.

Indeed, that doesn't work. The flag has been deprecated a while ago: https://github.com/jOOQ/jOOQ/issues/3770

The flag had been added very early in jOOQ 1.x, long before `Converter` and `Binding` types were available. It's usually better to implement the correct type yourself. Once you work with a `Converter` or `Binding`, you can use the usual methods to connect it to the relevant columns.

> The above is the closest I’ve come, but as you see it breaks down when parsing the incoming user object.

Well, you can always use the `ConverterProvider` API (which is used behind the scenes) to convert a `String` to a `LocalDate`...? Does that help?

**Comment by alf** [comment]
Are you saying that I should be able to implement the necessary handling of DATE columns by using a custom binding without setting the deprecated option? If so I must confess that I haven’t figured out how and would very much like to learn.

**Comment by lukaseder** [comment]
I'll post an example next week

**Comment by lukaseder** [comment]
I'm sorry for the delay. I had already closed this issue before, hence it got off my radar.

I'm curious about:

> I read in some thread in the mailing list that the deprecated option was left in because it wasn’t currently possible to replicate some aspect of this option, and my attempts at figuring it out failed miserably ☹️

Where did you read that? There's this blog post here: https://blog.jooq.org/2014/12/22/are-you-binding-your-oracle-dates-correctly-i-bet-you-arent/

That's about performance, not correctness, though. In any case, here are the steps:

1. You'll identify which columns are dates and which ones are timestamps first
2. You don't have to do anything about the "`DATE` as `DATE`" columns, because if you're using `java.sql.Date` or `java.time.LocalDate`, you'll get the correct behaviour and performance
3. You can use a data type `Binding` for the "`DATE` as `TIMESTAMP`", which implements a `CAST(? AS DATE)` whenever it binds a bind variable. An example implementation can be found at `org.jooq.impl.DateAsTimestampBinding`.

The documentation about data type bindings can be found here:
https://www.jooq.org/doc/latest/manual/sql-building/queryparts/custom-bindings/

Does this help?

**Comment by alf** [comment]
Indeed it does. I'll have a go. Thank you for following up. :)

If I find the reference that made me believe that the deprecated option was left in because it was still needed then I'll update.

**Comment by alf** [comment]
You refer to DateAsTimestampBinding as an example of implementing `cast(? as date)` when binding a bind variable. When reading the source of this class here on Github I see no such example. This was a bit confusing until I thought of checking the file in the source code that comes with our licensed copy. Now I see.

**Comment by lukaseder** [comment]
Yes, regrettably, Github doesn't have an easy option for licensed customers to look at the closed source code (I mean from an ACL perspective. It doesn't make sense to manually add each customer to the jOOQ organisation).

**Comment by alf** [comment]
Not a problem, it just wasn't obvious to me at first why I couldn't figure out why the code you referred to didn't do what you said. :facepalm:

**Comment by alf** [comment]
I'm afraid I'm not able to make this work. 

The jOOQ code generator sets the database type to `java.time.LocalDate`, so my binding must be a `Binding<LocalDate, OffsetDateTime>` for the code to compile. Since the Binding must implement a `Converter<LocalDate, OffsetDateTime> converter()` I'm at a loss. Maybe there is a way for me to override the database type? If I could set it to `LocalDateTime` or even `Timestamp` I'd be able to make this work.

The example you refer to uses the `Timestamp` database type. It can do this, since it is used when `dateAsTimestamp` is true, and thus the code generator uses that database types when encountering DATE.

**Comment by lukaseder** [comment]
> Since the Binding must implement a `Converter<LocalDate, OffsetDateTime> converter()` I'm at a loss.

You have to *declare* a converter, but your binding doesn't have to *use* the converter. There may be some cases where jOOQ's internals will use your converter, mostly to get a hold of the `Class<T>` and `Class<U>` references, not necessarily for data type conversion. jOOQ will pass the `U` type to your `Binding` letting you implement the conversion yourself. It's not very intuitive, see also https://github.com/jOOQ/jOOQ/issues/4819, which attempts to fix this unnecessary coupling of `Binding` and `Converter`

**Comment by alf-mindshift** [comment]
With that extra piece of understanding under my belt, I'm happy to say that I've got it working. Thank you for your patience.

**Comment by alf** [comment]
With that extra piece of understanding under my belt, I'm happy to say that I've got it working. Thank you for your patience.

**Comment by alf** [comment]
What threw me for a spin was:

1. dateAsTimestamp is deprecated, but jOOQ does not provide a replacement binding out of the box and I found no examples/documentation that led me by the hand like you just did. This made the transition harder than I expected.
2. The documentation for Binding vs Converter is very terse. I needed to dive in the deep end and understand what's really happening before I felt safe implementing this.

So yes, some  more prose regarding when a converter is sufficient and the relationship between the converter and binding / why a binding is more powerful etc. Examples are important for me to really understand.
